### DISCUSSION TITLE
List manipulation, e.g., removal of quickfix entries.

### DISCUSSION THREAD

**Discussion by gregorias** [folke/trouble.nvim#347] [General]
Is it possible to delete entries from an existing list like the quickfix list? With the native quickfix window, I was able to configure `dd` to remove entries like so:

```vim
" When using `dd` in the quickfix list, remove the item from the quickfix list.
function! quickfixdd#RemoveQFItem() abort
  let l:qf_list = getqflist()

  if len(l:qf_list) > 0
    let l:curqfidx = line('.') - 1
    call remove(l:qf_list, l:curqfidx)
    call setqflist(l:qf_list, 'r')
  endif

  if len(l:qf_list) > 0
    execute l:curqfidx + 1 . 'cfirst'
    copen
  else
    cclose
  endif
endfunction
```

It would be cool if Trouble could also do that.

**Comment by 73** [comment]
I just tried it and `dd` removes an entry.

**Reply by georgev93** [reply]
Does it actually remove it from the quickfix list, though? When I do it, it removes the line from the window (which effectively looks like it removed it from the quickfix list), but I get a red notification letting me know the refresh is now set to off.

The removal of the line lasts as long as I'm still showing the same trouble quickfix window. If I close and reopen the window, the item I tried to delete shows up on the window again. It looks as if `dd` removes the item from the buffer, but not the actual quickfix list.

**Reply by AlejandroSanchez90** [reply]
Did you manage to figure this out? the quickfix wont update after delete, if i hit dd, it will remove it from the ui, but if i hit r then it will come back, i tried toggling refresh, but when i try this again, i get the auto refresh is disabled

**Reply by PatrikHajek** [reply]
@AlejandroSanchez90, @georgev93 see my [answer](https://github.com/folke/trouble.nvim/discussions/347#discussioncomment-15891195) if you're still interested.

**Comment by PatrikHajek** [comment]
Don't know if this is still relevant / will help someone, but I managed to make it work today.

```lua
--- Delete items from Trouble quickfix window.
--- @param ids string[] List of item ids to be deleted.
local function delete(ids)
  local trouble = require 'trouble'

  --- @param item trouble.Item
  local items = vim.tbl_filter(function(item)
    return not vim.list_contains(ids, item.id)
  end, trouble.get_items())

  --- @param item trouble.Item
  --- @type vim.quickfix.entry[]
  local qf_entries = vim.tbl_map(function(item)
    return item.item
  end, items)
  vim.fn.setqflist(qf_entries, 'r')
  trouble.refresh()
end

-- Add these keys to the Trouble config options.
local keys = {
  ['d'] = {
    --- @param view trouble.View
    action = function(view)
      -- Update this condition if you use `qflist`.
      if view.opts.mode ~= 'quickfix' then
        print 'Deletions only work in quickfix windows!'
        return
      end

      local selection = view:selection()
      --- @type string[]
      local ids = {}
      for _, node in ipairs(selection) do
        --- @param item trouble.Item
        local item_ids = vim.tbl_map(function(item)
          return item.id
        end, node:flatten())
        vim.list_extend(ids, item_ids)
      end

      delete(ids)
    end,
    desc = 'Delete selected nodes',
  },
  ['dd'] = {
    --- @param view trouble.View
    action = function(view)
      -- Update this condition if you use `qflist`.
      if view.opts.mode ~= 'quickfix' then
        print 'Deletions only work in quickfix windows!'
        return
      end

      local at = view:at()
      if at.node ~= nil then
        --- @param item trouble.Item
        local item_ids = vim.tbl_map(function(item)
          return item.id
        end, at.node:flatten())
        delete(item_ids)
      else
        print "Couldn't get current node"
      end
    end,
    desc = 'Delete the node under the cursor',
  },
}
```

For the full trouble setup, see my [config](https://github.com/PatrikHajek/kickstart.nvim/blob/84cac4b47ef5ab07de46290d2c9588b6eb22f295/lua/custom/plugins/trouble.lua).

If you use `qflist` mode instead of `quickfix` (or both), change the condition at the beginning of the actions to include the desired modes. Keep in mind that the delete functionality only works with the quickfix list.

(Might not work properly with a filtered Trouble quickfix window, haven't tested that.)
